### ISSUE TITLE
FR: Fallback e-ink display refresh option

### ISSUE THREAD

**Issue by I2ujbNGakP8KfLKbho0Z** [koreader/koreader#11288]
**Does your feature request involve difficulty completing a task? Please describe.**

Android e-readers for which Koreader is not optimized, or which have drivers not yet implemented for Koreader, suffer from ghosting just like any other e-ink device. It is not very feasible to implement drivers for all of them as they are often closed source or have a small user base.

**Describe the solution you'd like**

An option to trigger a cycle of black and white full screen flashes (2 cycles should be enough) on every page turn (or as specified) on any device, without checking whether it is an e-ink device or not.

**Describe alternatives you've considered**

The e-ink options already available in the app for each device aren't really sufficient, because they only apply to menus when the device is not detected as an e-reader.

**Comment by mergen3107** [comment]
> to trigger a cycle of black and white full screen flashes

IIRC, driver support is needed to do _anything_ with eink refresh...

**Comment by NiLuJe** [comment]
The option to request a flash on every page turn already exists (and still makes sense even on modern devices (except Bellatrix Kindles, where you can instead opt for the page turn animation, which works surprisingly well for image content), for some use-cases (e.g., CBZ), depending on user preference).

I will hard veto anything crazier like double flashes or flash to white then flash or crap like that, so, err, no, thanks, get a device that actually works as expected instead ;).

**Comment by hugleo** [comment]
> An option to trigger a cycle of black and white full screen flashes (2 cycles should be enough) on every page turn (or as specified) on any device, without checking whether it is an e-ink device or not.

Chatgpt:

Certainly, let me break down the process of how an e-ink display achieves grayscale and why drawing two cycles of white and black isn't sufficient for creating a range of shades.

    Microcapsules and Particles:
        In an e-ink display, each pixel is made up of microcapsules containing tiny particles suspended in a clear fluid.
        These particles are typically charged positively (white) or negatively (black).

    Initial State:
        In the initial state, the particles are evenly distributed throughout the microcapsules, making the pixel appear white.
        Alternatively, the particles may be gathered at the bottom of the microcapsules, making the pixel appear black.

    Applying Electric Field:
        When an electric field is applied, it affects the charged particles within the microcapsules.
        The direction of the electric field determines whether the particles move to the top (making the pixel appear white) or to the bottom (making the pixel appear black).

    Gray Shades:
        To achieve gray shades, the e-ink display uses varying voltages to control the extent of particle movement within the microcapsules.
        By adjusting the voltage, it is possible to have the particles partially move, resulting in intermediate states between fully white and fully black. This ability to control the positioning of particles allows for grayscale representation.

    Full Refresh:
        When updating the content on an e-ink display, a full refresh involves changing the state of all microcapsules on the screen.
        This is necessary because, unlike traditional displays where pixels can change color rapidly, e-ink relies on the physical movement of particles, which takes more time.

    Limitations of Simplistic Approaches:
        Drawing only two cycles of white and black would result in a binary representation – either fully white or fully black – with no variation in shades between them.
        To display a photograph, text, or any content with gradients, the e-ink display must be capable of controlling the positioning of particles to create the desired grayscale levels.

In summary, the ability to manipulate particles within microcapsules with varying voltages allows e-ink displays to produce a broad range of grayscale shades. The full refresh involves updating the entire screen to reflect the desired content, taking advantage of the control over particle positioning to create detailed and varied images.

**Comment by hugleo** [comment]
IMHO I also believe that you can't beat the hardware Hz speed in modes other than fast. The flashes would be too slow.

**Comment by NiLuJe** [comment]
For context, there is *one* place where we do a "flash to white -> flash to image" dance, and that's the screensaver. It has... very little benefit in practic,e mainly because of the timing. You'd need to let the device "sit" on the white page for a good long while for it to make a moire noticeable difference.

So, doing something like that at runtime at a point where the device is *expected* to be responsive (unlike the screensaver): hell no.

**Comment by hugleo** [comment]
> For context, there is _one_ place where we do indeed do a "flash to white -> flash to image" dance (at least on _some_ platforms), and that's the screensaver. It has... very little benefit in practice, mainly because of the timing. You'd need to let the screen "sit" on the white page for a good long while for it to make a more noticeable difference.
> 
> So, doing something like that at runtime at a point where the device is _expected_ to be responsive (unlike the screensaver): hell no.

I also read somewhere that Onyx uses this black/white feature when pinching zooming the document

**Comment by NiLuJe** [comment]
> when pinching zooming

That sounds... funky.

I could imagine something like doing the pinch refreshes (if any) in A2 (... which might require further refresh trickery *before* engaging the first A2 refresh, though), and then a sane full-screen GC or GL or REAGL (flashing or not) once the zooming ends.

Oh, well ;D.

**Comment by pazos** [comment]
Another hard vetto here.

This app requires proper routines. Get a supported device or use another apk that can refresh the eink screen just by invalidating the parent view: (ie: Librera Reader https://github.com/foobnix/LibreraReader/issues/374)


> An option to trigger a cycle of black and white full screen flashes (2 cycles should be enough) on every page turn (or as specified) on any device, without checking whether it is an e-ink device or not.

I would rather prefeer to remove support for android completely

**Comment by Frenzie** [comment]
> > to trigger a cycle of black and white full screen flashes
> 
> IIRC, driver support is needed to do _anything_ with eink refresh...

Quite true, but someone sufficiently motivated might be able to hack something with the Android view model (assuming devices use that to trigger eink updates... maybe they don't, but I think some do anyway)

But that'd be kind of insane compared to the alternative of implementing a relevant driver.
